homeassistant/components/google_assistant/smart_home.py

```diff
@@ -133,9 +133,30 @@ def sync_serialize(self):
             device['name']['nicknames'] = aliases
 
         # add room hint if annotated
-        room = entity_config.get(CONF_ROOM_HINT)
-        if room:
-            device['roomHint'] = room
+        async def set_room():
+            room = entity_config.get(CONF_ROOM_HINT)
+            if room:
+                device['roomHint'] = room
+                return
+
+            dev_reg, ent_reg, area_reg = await gather(
+                self.hass.helpers.device_registry.async_get_registry(),
+                self.hass.helpers.entity_registry.async_get_registry(),
+                self.hass.helpers.area_registry.async_get_registry(),
+            )
+
+            entity_entry = ent_reg.async_get(state.entity_id)
+            if not (entity_entry and entity_entry.device_id):
+                return
+
+            device_entry = dev_reg.devices.get(entity_entry.device_id)
```

Comment:
I will suggest add a `device_registry.async_get()` method to replace `devices.get()`, and do the same enhancement in `area_registry` as well. Directly access `devices` and `areas` are not good design.

Moreover, it will be very helpful if we have a helper input is a entity_id, output are the entity_entry, device_entry, and area_entry

I actually have those in my another un-submitted pr.

Reply:
That is a good point. 

I've also been thinking about if we should get the registries squared out earlier and store it on `hass`, so we don't have to await whenever we use it. Downside is that we delay startup a little bit

Reply:
The startup delay isn't going to be that long, the asynchio thread is free to go do other stuff while it waits for the data to load from disk. I expect the registry loading is going to happen pretty early in the piece anyhow because at the very least the entity registry needs to load before a lot of the components, so it's not like you've really gained anything much by awaiting it every time you look for it.

I guess at minimum you have to load the config files and the three registries from disk because basically everything else hangs off these. Would an asyncio.gather hitting these bits of IO at once be worthwhile? Then you can just store the objects somewhere and not have to worry about awaiting any of their methods. 

OTOH, awaiting a method that doesn't actually do any IO isn't going to be very expensive so there's no major harm in occasional awaits, only it gets annoying because you have to propagate the async marker to everything that calls them.

Reply:
Some of these files can grow pretty big, there are people with 1500 entities. If we delay start up, we cannot initialize any component until all three are loaded. In the future more data might be added, and we end up delaying startup even more.

Reply:
I just did a test run with the most basic possible config, and ``entity_registry.async_get_registry``` and ```device_registry.async_get_registry``` does get called on start up, it's hit as soon as any platform calls the ```async_add_entities``` method in ```async_setup_platform``` for the component which for me happens with the yr platform. This hits ```entity_platform.async_add_entities```, so really no components can become available until at least the entity and device registries are loaded from disk.

Of course this is asynchronous, so we can get on and do other work in the mean time. Which is a reasonable enough argument to leave it how it is.

Reply:
Agree with the off-topic and good point on it being used inside Entity Platform, so we might as well make it available right away.